ArduCopter/rate_thread.cpp

```diff
@@ -0,0 +1,501 @@
+ Rate loop timing: gyro=177us, rate=13us, motors=46us, log=7us, ctrl=1us
+
+ The log output is an average since it only runs at 1Khz, so roughly 28us elapsed. So the majority of the time
+ is spent waiting for a gyro sample (higher is better here since it represents the idle time) updating the PIDs
+ and outputting to the motors. Everything else is relatively cheap. Since the total cycle time is 250us the duty
+ cycle is thus 29%
+ */
+
+#define DIV_ROUND_INT(x, d) ((x + d/2) / d)
+
+uint8_t Copter::calc_gyro_decimation(uint8_t gyro_decimation, uint16_t rate_hz)
+{
+    return MAX(uint8_t(DIV_ROUND_INT(ins.get_raw_gyro_rate_hz() / gyro_decimation, rate_hz)), 1U);
+}
+
+static inline bool run_decimated_callback(uint8_t decimation_rate, uint8_t& decimation_count)
+{
+    return decimation_rate > 0 && ++decimation_count >= decimation_rate;
+}
+
+//#define RATE_LOOP_TIMING_DEBUG
+/*
+  thread for rate control
+*/
+void Copter::rate_controller_thread()
+{
+    uint8_t target_rate_decimation = constrain_int16(g2.att_decimation.get(), 1, DIV_ROUND_INT(ins.get_raw_gyro_rate_hz(), 400));
+    uint8_t rate_decimation = target_rate_decimation;
+
+    // set up the decimation rates
+    RateControllerRates rates;
+    rate_controller_set_rates(rate_decimation, rates, false);
+
+    uint32_t rate_loop_count = 0;
+    uint32_t prev_loop_count = 0;
+
+    uint32_t last_run_us = AP_HAL::micros();
+    float max_dt = 0.0;
+    float min_dt = 1.0;
+    uint32_t now_ms = AP_HAL::millis();
+    uint32_t last_rate_check_ms = 0;
+    uint32_t last_rate_increase_ms = 0;
+#if HAL_LOGGING_ENABLED
+    uint32_t last_rtdt_log_ms = now_ms;
+#endif
+    uint32_t last_notch_sample_ms = now_ms;
+    bool was_using_rate_thread = false;
+    bool notify_fixed_rate_active = true;
+    bool was_armed = false;
+    uint32_t running_slow = 0;
+#ifdef RATE_LOOP_TIMING_DEBUG
+    uint32_t gyro_sample_time_us = 0;
+    uint32_t rate_controller_time_us = 0;
+    uint32_t motor_output_us = 0;
+    uint32_t log_output_us = 0;
+    uint32_t ctrl_output_us = 0;
+    uint32_t timing_count = 0;
+    uint32_t last_timing_msg_us = 0;
+#endif
+
+    // run the filters at half the gyro rate
+#if HAL_LOGGING_ENABLED
+    uint8_t log_loop_count = 0;
+#endif
+    uint8_t main_loop_count = 0;
+    uint8_t filter_loop_count = 0;
+
+    while (true) {
+
+#ifdef RATE_LOOP_TIMING_DEBUG
+        uint32_t rate_now_us = AP_HAL::micros();
+#endif
+
+        // allow changing option at runtime
+        if (get_fast_rate_type() == FastRateType::FAST_RATE_DISABLED || ap.motor_test) {
+            if (was_using_rate_thread) {
+                disable_fast_rate_loop(rates);
+                was_using_rate_thread = false;
+            }
+            hal.scheduler->delay_microseconds(500);
+            last_run_us = AP_HAL::micros();
+            continue;
+        }
+
+        // set up rate thread requirements
+        if (!using_rate_thread) {
+            enable_fast_rate_loop(rate_decimation, rates);
```

Comment:
not a blocker, feel free to resolve immediately but we need to be careful that we don't somehow end up in a loop where we continuously delete and re-create the buffer.

Reply:
Fixed